File: problems/023_non_abundant_sums/non_abundant_sums.py
# non_abundant_sums.py
# Python reference solution for Project Euler 23. Final answer redacted.

def spd_table(limit: int) -> list[int]:
    spd = [0] * (limit + 1)
    for i in range(1, limit // 2 + 1):
        for j in range(i * 2, limit + 1, i):
            spd[j] += i
    return spd

def main() -> None:
    LIMIT = 28123
    spd = spd_table(LIMIT)
    abundants = [n for n in range(1, LIMIT + 1) if spd[n] > n]

    can = [False] * (LIMIT + 1)
    for i, a in enumerate(abundants):
        for b in abundants[i:]:
            s = a + b
            if s > LIMIT:
                break
            can[s] = True

    total = sum(n for n in range(1, LIMIT + 1) if not can[n])
    # print(total)  # TODO: reveal locally
    print(f"Sum (<= {LIMIT}) of non-abundant-sum numbers = [redacted]")

if __name__ == "__main__":
    main()

File: problems/023_non_abundant_sums/non_abundant_sums.go
// non_abundant_sums.go
// Go reference solution for Project Euler 23. Final answer redacted.
package main

import "fmt"

func spdTable(limit int) []int {
    spd := make([]int, limit+1)
    for i := 1; i <= limit/2; i++ {
        for j := i * 2; j <= limit; j += i {
            spd[j] += i
        }
    }
    return spd
}

func main() {
    const LIMIT = 28123
    spd := spdTable(LIMIT)

    abundants := make([]int, 0)
    for n := 1; n <= LIMIT; n++ {
        if spd[n] > n {
            abundants = append(abundants, n)
        }
    }

    can := make([]bool, LIMIT+1)
    for i := 0; i < len(abundants); i++ {
        for j := i; j < len(abundants); j++ {
            s := abundants[i] + abundants[j]
            if s > LIMIT { break }
            can[s] = true
        }
    }

    var total int64 = 0
    for n := 1; n <= LIMIT; n++ {
        if !can[n] { total += int64(n) }
    }

    // fmt.Println(total) // TODO: reveal locally
    fmt.Printf("Sum (<= %d) of non-abundant-sum numbers = [redacted]\n", LIMIT)
}
